File: src/services/CategoryService.ts
import { pool } from '../config/database';
import { createError } from '../middleware/errorHandler';

interface Category {
  id: string;
  user_id: string;
  name: string;
  color: string;
  icon: string;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

interface CreateCategoryData {
  name: string;
  color: string;
  icon: string;
}

export class CategoryService {
  async getUserCategories(userId: string): Promise<Category[]> {
    const client = await pool.connect();
    try {
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      const query = `
        SELECT id, user_id, name, color, icon, is_default, created_at, updated_at
        FROM categories
        WHERE user_id = $1
        ORDER BY is_default DESC, name ASC
      `;
      
      const result = await client.query(query, [userId]);
      return result.rows;
    } catch (error) {
      console.error('Error retrieving categories:', error);
      throw createError('Failed to retrieve categories', 500);
    } finally {
      client.release();
    }
  }

  async createCategory(userId: string, data: CreateCategoryData): Promise<Category> {
    const client = await pool.connect();
    try {
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      // Check if category with same name already exists
      const existingQuery = `
        SELECT id FROM categories
        WHERE user_id = $1 AND LOWER(name) = LOWER($2)
      `;
      const existingResult = await client.query(existingQuery, [userId, data.name]);
      
      if (existingResult.rows.length > 0) {
        throw createError('Category with this name already exists', 400);
      }

      const query = `
        INSERT INTO categories (user_id, name, color, icon, is_default)
        VALUES ($1, $2, $3, $4, $5)
        RETURNING *
      `;
      
      const result = await client.query(query, [
        userId,
        data.name,
        data.color,
        data.icon,
        false
      ]);
      
      return result.rows[0];
    } catch (error: any) {
      console.error('Error creating category:', error);
      if (error.message && error.message.includes('already exists')) {
        throw error;
      }
      throw createError('Failed to create category', 500);
    } finally {
      client.release();
    }
  }

  async updateCategory(categoryId: string, userId: string, updateData: Partial<CreateCategoryData>): Promise<Category> {
    const client = await pool.connect();
    try {
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      // Check if category exists and belongs to user
      const checkQuery = `
        SELECT id FROM categories
        WHERE id = $1 AND user_id = $2
      `;
      const checkResult = await client.query(checkQuery, [categoryId, userId]);
      
      if (checkResult.rows.length === 0) {
        throw createError('Category not found', 404);
      }

      // If updating name, check for duplicates
      if (updateData.name) {
        const duplicateQuery = `
          SELECT id FROM categories
          WHERE user_id = $1 AND LOWER(name) = LOWER($2) AND id != $3
        `;
        const duplicateResult = await client.query(duplicateQuery, [userId, updateData.name, categoryId]);
        
        if (duplicateResult.rows.length > 0) {
          throw createError('Category with this name already exists', 400);
        }
      }

      // Build dynamic update query
      const updateFields = [];
      const updateValues = [];
      let paramCount = 1;

      if (updateData.name) {
        updateFields.push(`name = $${paramCount++}`);
        updateValues.push(updateData.name);
      }
      if (updateData.color) {
        updateFields.push(`color = $${paramCount++}`);
        updateValues.push(updateData.color);
      }
      if (updateData.icon) {
        updateFields.push(`icon = $${paramCount++}`);
        updateValues.push(updateData.icon);
      }

      if (updateFields.length === 0) {
        throw createError('No fields to update', 400);
      }

      updateFields.push(`updated_at = CURRENT_TIMESTAMP`);
      updateValues.push(categoryId, userId);

      const query = `
        UPDATE categories
        SET ${updateFields.join(', ')}
        WHERE id = $${paramCount++} AND user_id = $${paramCount++}
        RETURNING *
      `;
      
      const result = await client.query(query, updateValues);
      return result.rows[0];
    } catch (error: any) {
      console.error('Error updating category:', error);
      if (error.message && (error.message.includes('not found') || error.message.includes('already exists'))) {
        throw error;
      }
      throw createError('Failed to update category', 500);
    } finally {
      client.release();
    }
  }

  async deleteCategory(categoryId: string, userId: string): Promise<void> {
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      // Check if category exists and belongs to user
      const checkQuery = `
        SELECT id, is_default FROM categories
        WHERE id = $1 AND user_id = $2
      `;
      const checkResult = await client.query(checkQuery, [categoryId, userId]);
      
      if (checkResult.rows.length === 0) {
        throw createError('Category not found', 404);
      }

      const category = checkResult.rows[0];
      if (category.is_default) {
        throw createError('Cannot delete default categories', 400);
      }

      // Check if category is being used by transactions
      const transactionQuery = `
        SELECT COUNT(*) as count FROM transactions
        WHERE category_id = $1 AND user_id = $2
      `;
      const transactionResult = await client.query(transactionQuery, [categoryId, userId]);
      const transactionCount = parseInt(transactionResult.rows[0].count);

      if (transactionCount > 0) {
        throw createError('Cannot delete category that is being used by transactions', 400);
      }

      // Delete the category
      const deleteQuery = `
        DELETE FROM categories
        WHERE id = $1 AND user_id = $2
      `;
      await client.query(deleteQuery, [categoryId, userId]);

      await client.query('COMMIT');
    } catch (error: any) {
      await client.query('ROLLBACK');
      console.error('Error deleting category:', error);
      if (error.message && (error.message.includes('not found') || error.message.includes('Cannot delete'))) {
        throw error;
      }
      throw createError('Failed to delete category', 500);
    } finally {
      client.release();
    }
  }
}

File: src/services/category_service.py
from typing import List, Dict, Any, Optional
import uuid
from src.config.database import DatabaseManager
from src.models.category import CategoryCreate, CategoryUpdate
from src.middleware.error_handler import create_error
from src.utils.logger import logger


class CategoryService:
    """Service for category-related operations"""
    
    async def get_user_categories(self, user_id: str) -> List[Dict[str, Any]]:
        """Get all categories for a user"""
        try:
            query = """
                SELECT id, user_id, name, color, icon, is_default, created_at, updated_at
                FROM categories
                WHERE user_id = $1
                ORDER BY is_default DESC, name ASC
            """
            
            async with DatabaseManager(user_id) as db:
                rows = await db.fetch(query, uuid.UUID(user_id))
                categories = [dict(row) for row in rows]
                
                logger.info(f"Retrieved {len(categories)} categories for user {user_id}")
                return categories
        
        except Exception as e:
            logger.error(f"Error getting categories: {e}")
            raise create_error("Failed to retrieve categories", 500)
    
    async def create_category(self, user_id: str, category_data: CategoryCreate) -> Dict[str, Any]:
        """Create a new category"""
        try:
            async with DatabaseManager(user_id) as db:
                # Check if category with same name already exists
                existing_query = """
                    SELECT id FROM categories
                    WHERE user_id = $1 AND LOWER(name) = LOWER($2)
                """
                existing = await db.fetchrow(
                    existing_query, 
                    uuid.UUID(user_id), 
                    category_data.name
                )
                
                if existing:
                    raise create_error("Category with this name already exists", 400)
                
                # Create new category
                create_query = """
                    INSERT INTO categories (user_id, name, color, icon, is_default)
                    VALUES ($1, $2, $3, $4, $5)
                    RETURNING id, user_id, name, color, icon, is_default, created_at, updated_at
                """
                
                category = await db.fetchrow(
                    create_query,
                    uuid.UUID(user_id),
                    category_data.name,
                    category_data.color,
                    category_data.icon,
                    False
                )
                
                logger.info(f"Category created successfully: {category['id']}")
                return dict(category)
        
        except Exception as e:
            logger.error(f"Error creating category: {e}")
            if "already exists" in str(e):
                raise e
            raise create_error("Failed to create category", 500)
    
    async def update_category(
        self, 
        category_id: str, 
        user_id: str, 
        category_data: CategoryUpdate
    ) -> Optional[Dict[str, Any]]:
        """Update a category"""
        try:
            async with DatabaseManager(user_id) as db:
                # Check if category exists and belongs to user
                check_query = """
                    SELECT id FROM categories
                    WHERE id = $1 AND user_id = $2
                """
                existing = await db.fetchrow(
                    check_query, 
                    uuid.UUID(category_id), 
                    uuid.UUID(user_id)
                )
                
                if not existing:
                    raise create_error("Category not found", 404)
                
                # If updating name, check for duplicates
                if category_data.name:
                    duplicate_query = """
                        SELECT id FROM categories
                        WHERE user_id = $1 AND LOWER(name) = LOWER($2) AND id != $3
                    """
                    duplicate = await db.fetchrow(
                        duplicate_query,
                        uuid.UUID(user_id),
                        category_data.name,
                        uuid.UUID(category_id)
                    )
                    
                    if duplicate:
                        raise create_error("Category with this name already exists", 400)
                
                # Build dynamic update query
                update_fields = []
                params = []
                param_count = 1
                
                update_data = category_data.dict(exclude_unset=True)
                
                for field, value in update_data.items():
                    if value is not None:
                        update_fields.append(f"{field} = ${param_count}")
                        params.append(value)
                        param_count += 1
                
                if not update_fields:
                    raise create_error("No fields to update", 400)
                
                update_fields.append("updated_at = CURRENT_TIMESTAMP")
                
                # Add where clause parameters
                params.extend([uuid.UUID(category_id), uuid.UUID(user_id)])
                
                query = f"""
                    UPDATE categories
                    SET {', '.join(update_fields)}
                    WHERE id = ${param_count} AND user_id = ${param_count + 1}
                    RETURNING id, user_id, name, color, icon, is_default, created_at, updated_at
                """
                
                category = await db.fetchrow(query, *params)
                
                if not category:
                    raise create_error("Category not found", 404)
                
                logger.info(f"Category updated successfully: {category_id}")
                return dict(category)
        
        except Exception as e:
            logger.error(f"Error updating category: {e}")
            if "not found" in str(e) or "already exists" in str(e):
                raise e
            raise create_error("Failed to update category", 500)
    
    async def delete_category(self, category_id: str, user_id: str) -> bool:
        """Delete a category"""
        try:
            async with DatabaseManager(user_id) as db:
                async with db.transaction():
                    # Check if category exists and belongs to user
                    check_query = """
                        SELECT id, is_default FROM categories
                        WHERE id = $1 AND user_id = $2
                    """
                    category = await db.fetchrow(
                        check_query, 
                        uuid.UUID(category_id), 
                        uuid.UUID(user_id)
                    )
                    
                    if not category:
                        raise create_error("Category not found", 404)
                    
                    if category["is_default"]:
                        raise create_error("Cannot delete default categories", 400)
                    
                    # Check if category is being used by transactions
                    transaction_query = """
                        SELECT COUNT(*) as count FROM transactions
                        WHERE category_id = $1 AND user_id = $2
                    """
                    transaction_result = await db.fetchrow(
                        transaction_query, 
                        uuid.UUID(category_id), 
                        uuid.UUID(user_id)
                    )
                    
                    if transaction_result["count"] > 0:
                        raise create_error(
                            "Cannot delete category that is being used by transactions", 
                            400
                        )
                    
                    # Delete the category
                    delete_query = """
                        DELETE FROM categories
                        WHERE id = $1 AND user_id = $2
                    """
                    result = await db.execute(
                        delete_query, 
                        uuid.UUID(category_id), 
                        uuid.UUID(user_id)
                    )
                    
                    success = result.split()[-1] == "1"
                    
                    if success:
                        logger.info(f"Category deleted successfully: {category_id}")
                    
                    return success
        
        except Exception as e:
            logger.error(f"Error deleting category: {e}")
            if "not found" in str(e) or "Cannot delete" in str(e):
                raise e
            raise create_error("Failed to delete category", 500)